pkg/reconciler/controllers/remediator.go

```diff
@@ -0,0 +1,335 @@
+		// is for an Object that was deleted.
+		toRemediate = nil
+		// If the object was deleted, it has no ResourceVersion, use the constant.
+		currentRV = cache.DeletedResourceVersion
+	} else {
+		toRemediate = obj
+	}
+
+	// If the current ResourceVersion was last remediated, skip remediation because
+	// it should match the declared resource.
+	if currentRV == lastRV {
+		// If no fight errors, return directly and remove the remediate cache
+		if len(r.FightErrors()) == 0 {
+			r.RemoveObj(id)
+			return ctrl.Result{}, nil
+		}
+		// Even though remediation is skipped, it still needs to detect fight to
+		// report the fight status.
+		_, fightErr := r.DetectFight(time.Now(), obj)
+		if fightErr == nil {
+			// No more fight for the object, remove it from the queue.
+			r.RemoveFightError(id)
+			r.RemoveObj(id)
+			return ctrl.Result{}, nil
+		}
+
+		r.AddFightError(id, fightErr)
+		// Requeue the request to address the fight.
+		// Don't mark object done as it requeues the error directly
+		return ctrl.Result{}, fmt.Errorf("fight for %s still exists in the remediator", id)
+	}
+
+	// If the current ResourceVersion doesn't match the last remediated one, it is
+	// a new remediation.
+	klog.V(3).Infof("Remediator processing object %q (generation: %d)", id, obj.GetGeneration())
+	_, remErr := r.remediate(r.ControllerCtx, id, toRemediate)
+	if remErr != nil {
+		// To debug the set of events we've missed, you may need to comment out this
+		// block. Specifically, this makes things smooth for production, but can
+		// hide bugs (for example, if we don't properly process delete events).
+		if remErr.Code() == syncerclient.ResourceConflictCode {
+			// This means our cached version of the object isn't the same as the one
+			// on the cluster. We need to refresh the cached version.
+			if refreshErr := r.refresh(r.ControllerCtx, obj); refreshErr != nil {
+				klog.Errorf("Unable to update cached version of %q: %v", id, refreshErr)
+			}
+		}
+		klog.Errorf("Failed to remediate %q: %v", id, remErr)
+		// Requeue the remediate error for retry.
+		return ctrl.Result{}, remErr
+	}
+
+	klog.V(3).Infof("Successfully reconciled %q", id)
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager registers the Remediator Controller.
+func (r *Remediator) SetupWithManager(mgr ctrl.Manager) error {
+	// an event channel to trigger reconciliation
+	timerEvents := make(chan event.GenericEvent)
+
+	// run the timer as a goroutine to trigger the reconciliation periodically
+	go func() {
+		remediateTimer := time.NewTimer(time.Second)
+		defer remediateTimer.Stop()
+
+		for {
+			select {
+			case <-r.ControllerCtx.Done():
+				// The controller is shutting down, close the timerEvents channel and exit
+				close(timerEvents)
+				close(r.DoneCh) // inform finalizer the remediator is done
+				return
+
+			case <-remediateTimer.C:
+				if r.CanRemediate() {
+					// Skip enqueuing remediator requests if the Applier is running.
+					r.enqueueEvents(timerEvents)
```

Comment:
Since we are enqueueing objects for remediation, the following scenario seems possible:
- sync to commit A
- enqueue remediation requests for commit A
- new commit B is created, pausing remediator
- commit B is synced, and the inventory differs substantially from commit B
- remediator continues and beings processing remediation requests for commit A

In this hypothetical, would the remediator try to correct for drift based on the previous commit? If so, what would that look like? I'm thinking we may want to reset the queue after the syncer runs the applier

Reply:
If the above case happens, the remediator is basically a no-op for queued objects from commit A.
Here are the possible scenarios:
1. new commit B doesn't have the same GVK, so the watch is stopped. No new events will be added to the queue. Remaining items in the queue will be drained as a DELETE operation (the object is already deleted by the syncer for commit B, so no-op). 
2. new commit has the same GVK, but the object is updated. When commit B is synced, the watch receives the update event, updates the queue, and marks canRemediate as true. The remediator processes the request. The object in the queue is the same as the actual object, so no op.
3. new commit has the same GVK, and the object is the same. The watch won't receive any new event, so no remediation.
4. new commit has new GVK, new watch will be set up to trigger new remediator.

If we reset the queue after the syncer runs the applier, it helps avoid draining the queue in the first scenario only, which is still beneficial, but it requires all Remediate requests to be in the queue, which also means we can't use the controller-runtime to requeue the requests.

Reply:
done